### PR TITLE
Fix jnp.partition, jnp.argpartition and jnp.top_k on signed integer minimums

### PR DESCRIPTION
**Pull request by teddytennant**
## What

`jnp.partition`, `jnp.argpartition` and `jnp.top_k(mode='smallest')` now handle the smallest representable signed integer, and `partition`/`argpartition` accept boolean arrays as `top_k` and NumPy already do.

## Why

All three find the smallest entries by negating the input, calling `lax.top_k`, and negating back. For a signed integer `-INT_MIN` is `INT_MIN`, so the minimum compares as the maximum and never gets selected:

```python
>>> jnp.partition(jnp.array([-2**31, 5, 3], jnp.int32), 1)
Array([3, 5, 5], dtype=int32)   # np.partition gives [-2147483648, 3, 5]
```

That result is not a permutation of the input. The minimum is gone and `5` appears twice, which also breaks the pivot guarantee in the docstring.

The unsigned form of this is #22137, fixed in #22288 by special-casing `-(arr + 1)`, and `top_k` later grew a second special case for `bool`. Both of those are bitwise inversion written longhand, which is exactly what the signed case needs, so this replaces all three call sites with one `_order_reversing_invert` helper: `bitwise_not` for bool and every integer width, `neg` for floats. It is an involution, it is order-reversing, and it cannot overflow. Net effect is one fewer branch in each function and bool support in `partition`/`argpartition` for free. Float behavior is untouched.

## Testing

`testPartition` and `testArgpartition` draw from `jtu.rand_default`, which never emits `INT_MIN`, and the existing extreme-value tests are unsigned-only, so nothing covered this. Added `testPartitionSignedWithMinValue`, `testArgpartitionSignedWithMinValue`, `testPartitionBool`, `testArgpartitionBool` and `testTopKSmallestWithMinValue` alongside the unsigned versions; 25 of the new cases fail on `main`. I also ran a differential check against NumPy over 21130 `partition`/`argpartition`/`top_k` calls spanning every integer width, both float widths and bool, drawing values from the dtype extremes, with no mismatches. Full `tests/lax_numpy_test.py` passes (1167). CHANGELOG entry included.

### CODE REVIEW

File: jax/_src/numpy/sorting.py [hunk 1]
```diff
@@ -27,6 +27,22 @@
 
 export = set_module('jax.numpy')
 
+def _order_reversing_invert(x: Array) -> Array:
```
Comment (jakevdp) [inline jax/_src/numpy/sorting.py:30]: I think this utility claims too much: in particular for floating point, neg is not an order-reversing inversion, because NaN values are sorted to the end regardless of their sign.

But for bool and integer types, the approach seems sound.

A five-line paragraph in the docstring seems excessive here; I'd probably write something like this:
```python
def _order_reversing_invert(x: Array) -> Array:
   """Map an array to one whose sort order is reversed."""
  if dtypes.isdtype(x.dtype, ("bool", "integral")):
    # For boolean and integer types, bitwise_not is order-reversing over the full domain
    return lax.bitwise_not(x)
  else:
    # For inexact types, negation is order-reversing except in a few cases:
    # - NaN, which lax.sort places at the end regardless of sign
    # - signed/unsigned zero, which are always left in stable order regardless of sign.
    # In places where this is important, these must be handled specially below.
    return lax.neg(x)
```

File: tests/lax_numpy_test.py [hunk 2]
```diff
@@ -4499,6 +4499,32 @@ def testPartitionUnsignedWithZeros(self, kth, dtype):
     np_output = np.partition(arg, axis=axis, kth=kth)
     self._assertSamePartitionedArrays(jnp_output, np_output, axis, kth, shape)
 
+  @jtu.sample_product(
+    kth=range(10),
+    dtype=int_dtypes,
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: I think it wouldn't hurt to cover both `int_dtypes` and `unsigned_dtypes` here.

Also, we don't need a full range of k values for this test: I think something like `k=[-2, 2]` would be sufficient test coverage for the issue.

File: tests/lax_numpy_test.py [hunk 3]
```diff
@@ -4499,6 +4499,32 @@ def testPartitionUnsignedWithZeros(self, kth, dtype):
...
+    max_val = np.iinfo(dtype).max
+    arg = jnp.array([[6, max_val, min_val, 4, 3, 1, min_val, 7, 5, 2]], dtype=dtype)
+    axis = -1
+    shape = arg.shape
+    jnp_output = jnp.partition(arg, axis=axis, kth=kth)
+    np_output = np.partition(arg, axis=axis, kth=kth)
+    self._assertSamePartitionedArrays(jnp_output, np_output, axis, kth, shape)
+
+  @jtu.sample_product(
+    kth=range(6),
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: `kth=[-2, 2]` is probably sufficient

File: tests/lax_numpy_test.py [hunk 4]
```diff
@@ -4556,6 +4582,54 @@ def testArgpartitionUnsignedWithZeros(self, kth, dtype):
     np_values = getvals(arg, np_output)
     self._assertSamePartitionedArrays(jnp_values, np_values, axis, kth, shape)
 
+  @jtu.sample_product(
+    kth=range(10),
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: `k=[-2, 2]`

File: tests/lax_numpy_test.py [hunk 5]
```diff
@@ -4556,6 +4582,54 @@ def testArgpartitionUnsignedWithZeros(self, kth, dtype):
...
+    dtype=int_dtypes,
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: `dtype=int_dtypes + unsigned_dtypes`

Also adjust the test name to fit this.

File: tests/lax_numpy_test.py [hunk 6]
```diff
@@ -4556,6 +4582,54 @@ def testArgpartitionUnsignedWithZeros(self, kth, dtype):
...
+    getvals = lambda x, ind: x[ind]
+    for ax in range(arg.ndim):
+      if ax != range(arg.ndim)[axis]:
+        getvals = jax.vmap(getvals, in_axes=ax, out_axes=ax)
+    jnp_values = getvals(arg, jnp_output)
+    np_values = getvals(arg, np_output)
+    self._assertSamePartitionedArrays(jnp_values, np_values, axis, kth, shape)
+
+  @jtu.sample_product(
+    kth=range(6),
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: `kth=[-2, 2]`

File: tests/lax_numpy_test.py [hunk 7]
```diff
@@ -4556,6 +4582,54 @@ def testArgpartitionUnsignedWithZeros(self, kth, dtype):
...
+    dtype=bool_dtypes,
+  )
+  def testArgpartitionBool(self, kth, dtype):
+    arg = jnp.array([[True, False, True, True, False, False]], dtype=dtype)
+    axis = -1
+    shape = arg.shape
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: nit: delete this and use `arg.shape` directly in line 4631

File: tests/lax_numpy_test.py [hunk 8]
```diff
@@ -4556,6 +4582,54 @@ def testArgpartitionUnsignedWithZeros(self, kth, dtype):
...
+
+    # Assert that all indices are present
+    self.assertArraysEqual(jnp.sort(jnp_output, axis), np.sort(np_output, axis), check_dtypes=False)
+
+    # Because JAX & numpy may treat duplicates differently, we must compare values
+    # rather than indices.
+    getvals = lambda x, ind: x[ind]
+    for ax in range(arg.ndim):
+      if ax != range(arg.ndim)[axis]:
+        getvals = jax.vmap(getvals, in_axes=ax, out_axes=ax)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: since we know `arg` is one-dimensional and `axis=-1`, we could use `jnp_values = jnp_output[ind]` and delete all this extra logic.

Same in ine 4629

**Review by jakevdp — COMMENTED** [review]
Thanks for the fix! A couple comments below.

File: jax/_src/numpy/sorting.py (hunk 1 above)
Comment (teddytennant) [reply jax/_src/numpy/sorting.py:30]: used this

File: tests/lax_numpy_test.py (hunk 2 above)
Comment (teddytennant) [reply tests/lax_numpy_test.py]: int_dtypes + unsigned_dtypes, kth=[-2, 2]. renamed to testPartitionIntegerExtrema

File: tests/lax_numpy_test.py (hunk 3 above)
Comment (teddytennant) [reply tests/lax_numpy_test.py]: done

File: tests/lax_numpy_test.py (hunk 4 above)
Comment (teddytennant) [reply tests/lax_numpy_test.py]: kth=[-2, 2]

File: tests/lax_numpy_test.py (hunk 5 above)
Comment (teddytennant) [reply tests/lax_numpy_test.py]: renamed to testArgpartitionIntegerExtrema

File: tests/lax_numpy_test.py (hunk 6 above)
Comment (teddytennant) [reply tests/lax_numpy_test.py]: done

File: tests/lax_numpy_test.py (hunk 7 above)
Comment (teddytennant) [reply tests/lax_numpy_test.py]: dropped it

File: tests/lax_numpy_test.py (hunk 8 above)
Comment (teddytennant) [reply tests/lax_numpy_test.py]: flattened the fixture and gather with arg[indices]

File: tests/lax_numpy_test.py [hunk 9]
```diff
@@ -4499,6 +4499,32 @@ def testPartitionUnsignedWithZeros(self, kth, dtype):
...
+    shape = arg.shape
+    jnp_output = jnp.partition(arg, axis=axis, kth=kth)
+    np_output = np.partition(arg, axis=axis, kth=kth)
+    self._assertSamePartitionedArrays(jnp_output, np_output, axis, kth, shape)
+
+  @jtu.sample_product(
+    kth=[-2, 2],
+    dtype=bool_dtypes,
+  )
+  def testPartitionBool(self, kth, dtype):
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: It seems like we could delete `testPartitionBool` and `testArgPartitionBool` if we just included `bool_dtypes` in the parameterization of `testPartition` and `testArgPartition`

**Review by jakevdp — COMMENTED** [review]
Looks great! Just one last comment

File: tests/lax_numpy_test.py (hunk 9 above)
Comment (teddytennant) [reply tests/lax_numpy_test.py]: folded bool into the product and dropped the dedicated tests.

**Comment by teddytennant** [comment]
the layout_test.host_auto_layout and sinc failures are the same ones currently red on main.

**Review by jakevdp — APPROVED** [review]
Looks great – thanks!
